devicemapper: Check loop devices of existing pool

rhvgoyal opened the pull request:
Fixes #14020 

Often it happens that docker is not able to shutdown/remove the thin
pool it created because some device has leaked into some mount name
space. That means device is in use and that means pool can't be removed.

Docker will leave pool as it is and exit. Later when user starts the
docker, it finds pool is already there and docker uses it. But docker
does not know it is same pool which is using the loop devices. Now
docker thinks loop devices are not being used. That means it does not
display the data correctly in "docker info", giving user wrong information.

This patch tries to detect if loop devices as created by docker are
being used for pool and fills in the right details in "docker info".

Signed-off-by: Vivek Goyal vgoyal@redhat.com

rhvgoyal commented:
@vbatts Please have a look.

rhvgoyal commented:
@rhatdan This should fix the issue you were facing some time back when you killed the docker daemon using (Ctrl-C) and restarted and now docker info does not show loop devices.

vbatts commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -1083,6 +1083,126 @@ func determineDriverCapabilities(version string) error {
 }
 
+// Determine the major and minor number of loopback device
+func getLoopMajorMinor(file *os.File) (uint64, uint64, error) {
```
as for a function name, this function is not confined to loop files...

rhvgoyal replied:
Agreed. I think it should work on any block device. I will modify it.

rhvgoyal commented:
@vbatts I changed function name and refreshed PR. PTAL.

vbatts commented:
@rhvgoyal can you rebase on master?

rhvgoyal commented:
@vbatts rebased on latest master.

LK4D4 commented:
ping @vbatts 
because I have no idea :)

rhvgoyal commented:
@vbatts ping. Can you please review this one.

vbatts commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -1083,6 +1083,126 @@ func determineDriverCapabilities(version string) error {
+		return 0, 0, err
+	}
+
+	Dev := stat.Sys().(*syscall.Stat_t).Rdev
```
doesn't _really_ matter, but stylistically this `Dev` ought to be lower case.

rhvgoyal replied:
Will do.

vbatts commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -1083,6 +1083,126 @@ func determineDriverCapabilities(version string) error {
+	}
+
+	defer file.Close()
+	LoopbackDevice := devicemapper.FindLoopDeviceFor(file)
```
lower case

vbatts commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -1083,6 +1083,126 @@ func determineDriverCapabilities(version string) error {
+		return 0, 0, 0, 0, err
+	}
+
+	logrus.Debugf("devmapper: poolDataMajMin=%s poolMetaMajMin=%s\n", poolDataMajMin, poolMetadataMajMin)
```
for consistency sake, can this log prefix be `[devmapper]` instead of `devmapper:`

vbatts commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -1083,6 +1083,126 @@ func determineDriverCapabilities(version string) error {
+	defer file.Close()
+	LoopbackDevice := devicemapper.FindLoopDeviceFor(file)
+	if LoopbackDevice == nil {
+		return "", 0, 0, fmt.Errorf("devmapper: Unable to find loopback mount for: %s", filename)
```
log prefix

rhvgoyal replied:
I see many location in code where either we are not prefixing anything or just doing devmapper. I am not sure what's the convention now. Though it would be nice to make all of these uniform.

vbatts commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -1083,6 +1083,126 @@ func determineDriverCapabilities(version string) error {
+	return poolDataMajor, poolDataMinor, poolMetadataMajor, poolMetadataMinor, nil
+}
+
+func (devices *DeviceSet) loadThinPoolProperties() error {
```
this will only get called in a loop back scenario right? and the call to `devices.loopbackdDir()` assumes so. Maybe the function name could be more clear in that regard?

vbatts commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -1230,6 +1350,17 @@ func (devices *DeviceSet) initDevmapper(doInit bool) error {
+	// we probably created pool earlier and could not remove it as some
+	// containers were still using it. Detect some of the properties of
+	// pool, like is it using loop devices.
+	if info.Exists != 0 && devices.thinPoolDevice == "" {
```
is there no reason to get the major and minor info when `devices.thinPoolDevice` is already set?

rhvgoyal commented:
@vbatts 

I pushed my patches again after modifications "git push -f" and that seems to have resulted in closing this issue. I have no clue what happened here.

rhvgoyal committed:
devicemapper: Check loop devices of existing pool

Often it happens that docker is not able to shutdown/remove the thin
pool it created because some device has leaked into some mount name
space. That means device is in use and that means pool can't be removed.

Docker will leave pool as it is and exit. Later when user starts the
docker, it finds pool is already there and docker uses it. But docker
does not know it is same pool which is using the loop devices. Now
docker thinks loop devices are not being used. That means it does not
display the data correctly in "docker info", giving user wrong information.

This patch tries to detect if loop devices as created by docker are
being used for pool and fills in the right details in "docker info".

Signed-off-by: Vivek Goyal <vgoyal@redhat.com>

rhvgoyal commented:
I think I had pushed without my patch and that caused this. Now I have pushed with my commit again.

rhvgoyal commented:
@vbatts 

I have taken care of your review comments and pushed patches again. PTAL.

vbatts commented:
LGTM

vbatts commented:
@calavera care to review?

jfrazelle commented:
@vbatts i think this is ok to merge if its still alright with you, LGTM
